# Cry phylogeny
The repository with working codes and supplementary materials for the study of Cry toxins evolution

## Contents 

This repository contains scripts used for statistical processing, phylogeny reconstruction, inferencing recombination events, and analyzing evolutionary selection in the sequences of 3-domain Cry toxins. Please consult the Methods section in the paper for extra details:

## Figures
Figures from the main text are available in the `pics/` directory. For the description, please consult the Results section of the article.

## Supplementary
All supplementary material is located in the `supplementary/` directory. 
`Supplementary_figures.docx` is a Microsoft Word document with a description of all supplementary figures.
`Supplementary_tables.xlsx` is an Exel table with all supplementary tables.

* The `supplementary/pics/` folder contains individual supplementary figures.
* For convenience, in the `supplementary/tables/` folder individual supporting tables in CSV format are provided. For a detailed description of the tables, please consult the `Supplementary_tables.xlsx` file.

## Data
Analyzed data are included in the `data/` directory. The directory contains sequences and phylogenetic trees. Other essential tables obtained when using data processing scripts are presented in the `data_for_scripts/` directory. 

### Sequences
The `data/fastas` directory contains nucleotide and protein sequences, both full and domain-wise.
The raw set of 3-D Cry protein sequences is provided in the `source_Cry_proteins.fasta` file. The file contains full-length protein sequences inferred from four datasets: BPPRC (Bacterial Pesticidal Protein Resource Center), NCBI IPG database (<i>Bacillus thuringiensis</i>), 467 <i>Bt</i> assemblies from NCBI Assembly, and NCBI Genbank (bacteria non-redundant). Three-domain sequences were obtained with CryProcessor in the “<i>-do</i>” mode. The presented set of sequences is redundant and was further deduplicated.
* The directory `data/fastas/all_sequences` includes non-redundant set from the `source_Cry_proteins.fasta` file. The duplicates with a 100% identity threshold are discarded. Cry toxins absent in the natural strains (presented only in patents) and subjected to artificial mutagenesis are also excluded;</li>
* The directory `data/fastas/ref_sequences` includes reference sequences obtained by applying CD-HIT software with a 95% identity threshold on the non-redundant filtered set of sequences in the `all_sequences` directory.
All the files in two subdirectories follow the same rules for prefixes and postfixes:
* The `nucl_` prefix corresponds to nucleotide sequences;
* The `prot_` prefix marks protein sequences;
* Sequences of individual domains are marked with the `_domain` postfix;
* The `_full` postfix indicates that the sequences contain all regions, including N- and C-terminal ones;
* The `_ processed` postfix indicates that the sequences contain only three main functional domains and are cropped from the beginning of the first domain to the end of the third domain accordingly. 

### Phylogenetic trees
The `data/phylogeny` directory includes reconstructed phylogenies and multiple sequence alignments used for building phylogenetic trees. The directory contains the following subdirectories and files:
1. The `phylogeny/data` subdirectory comprises source files used for phylogeny reconstruction:
* The `phylogeny/data/concat_msa.fasta` file includes concatenated domain-wise alignments;
* The `phylogeny/data/concat_msa_trimmed.fasta` file is the source data for reconstructing reference phylogenetic tree derived from `concat_msa.fasta` using the trimAl utility;
* The `phylogeny/data/concat_msa.fasta` file is a set of partitions for trimmed sequences of the individual domains within the concatenated alignment;
* The `phylogeny/data/domains_msa` subdirectory includes both raw and trimmed alignments of the nucleotide sequences of individual domains;
2. The `phylogeny/init_trees` directory contains both domain-wise trees and the tree inferred from the concatenated partitioned alignment;
3. The `phylogeny/collapsed_trees` comprises domain-wise phylogenetic trees corrected for recombination events with parents and recombinants grouped in case of multifurcation. A detailed description of the filtering procedure is presented in the Methods section of the article.

## Data for scripts
For better reproducibility, the data needed for obtaining all the results described in the article is provided in the `data_for_scripts/` directory. The subdirectories are divided according to the subsections in the Results section of the article. The directory `data_for_scripts/for_viz` contains source data required for producing figures in the main text and supplementary material. The presented data could be generated by scripts in the `scripts/` directory. 

## Scripts
The `scripts/` directory includes all code used for pangenome analysis. The scrips are grouped into two categories, namely, those used for data processing and visualization. These groups are further subdivided following the subsections in the Results section of the article. For convenience, required data are presented in the `data_for_script /` directory. Here, commands to run the scripts are presented. The description of input files is given above. For convenience, the data presented in the `data_for_scripts/` is also divided into subdirectories according to sections in the manuscript. The paths for the input files are given relative to the `scripts/` directory.
